wagtail/snippets/bulk_actions/delete.py

```diff
@@ -1,19 +1,51 @@
+from django.utils.functional import cached_property
 from django.utils.text import capfirst
 from django.utils.translation import gettext_lazy as _
 from django.utils.translation import ngettext
 
+from wagtail.admin.views.generic import BeforeAfterHookMixin
 from wagtail.snippets.bulk_actions.snippet_bulk_action import SnippetBulkAction
 from wagtail.snippets.permissions import get_permission_name
 
 
-class DeleteBulkAction(SnippetBulkAction):
+class DeleteBulkAction(BeforeAfterHookMixin, SnippetBulkAction):
     display_name = _("Delete")
     action_type = "delete"
     aria_label = _("Delete selected snippets")
     template_name = "wagtailsnippets/bulk_actions/confirm_bulk_delete.html"
     action_priority = 30
     classes = {"serious"}
 
+    @cached_property
+    def _actionable_objects(self):
+        # Cache the actionable objects so that we can use them early in the
+        # dispatch() method, and reuse it later when needed. We don't name this
+        # property "actionable_objects" because the method returns a tuple of
+        # `(objects, {"items_with_no_access": items_with_no_access})` and the base
+        # class assigns the `objects` to self.actionable_objects.
+        return super().get_actionable_objects()
+
+    def get_actionable_objects(self):
+        return self._actionable_objects
+
+    # Run the snippet deletion hooks that we've had since before bulk actions
+    # were introduced.
+    # These are different from the base bulk action hooks that are run by
+    # __run_{before,after}_hooks(), as those are only called in form_valid().
+    # Here, we make use of the BeforeAfterHookMixin to run the before hook at
+    # the beginning of dispatch(), and the after hook at the end of form_valid(),
+    # which is how the delete-multiple view for snippets used to work.
+
+    def run_before_hook(self):
+        # The method returns a tuple of
+        # `(objects, {"items_with_no_access": items_with_no_access})`, hence [0]
+        objects = self.get_actionable_objects()[0]
```

Comment:
Could also be `self._actionable_objects[0]`, but that would bypass `get_actionable_objects()` if it has been overridden.